Add `OutlinesModel` to run Transformers, Llama.cpp, MLXLM, SGLang and vLLM via Outlines

RobinPicard opened the pull request:
This PR is a minimal proposition of an interface for the integration of Outlines in Pydantic AI.

The idea is that there would be single `OutlinesModel` that would take as an argument in its initialization an Outlines model instance and that would then be in charge of interacting with its interface. The class has class methods (`transformers`, `llama_cpp`...) to avoid having to first create the Outlines model oneself and instead directly providing the arguments required for each model.

The proposal only supports json schema as an output type as Pydantic AI is currently focused on that type.

RobinPicard commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,266 @@
+    ) from _import_error
+
+
+@dataclass
+class OutlinesStreamedResponse(StreamedResponse):
+    """Implementation of `StreamedResponse` for Outlines models."""
+
+    _model_name: str
+    _response: AsyncIterable[str]
+    _timestamp: datetime
+
+    async def _get_event_iterator(self) -> AsyncIterator[ModelResponseStreamEvent]:
+        async for event in self._response:
+            event = self._parts_manager.handle_text_delta(vendor_part_id='content', content=event)
+            if event is not None:  # pragma: no branch
+                yield event
+
+    @property
+    def model_name(self) -> str:
+        """Get the model name of the response."""
+        return self._model_name
+
+    @property
+    def timestamp(self) -> datetime:
+        """Get the timestamp of the response."""
+        return self._timestamp
```
I don't get it, sorry

RobinPicard commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,266 @@
+        model: OutlinesBaseModel | OutlinesAsyncBaseModel,
+        model_name: str | None = None,
+        *,
+        provider: Literal['outlines'] | Provider[OutlinesBaseModel] = 'outlines',
```
How so?

DouweM commented on outlines_example.py:
We'll want to remove this before we merge and move it to docs

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,352 @@
+            or model_request_parameters.output_tools
+        ):
+            raise NotImplementedError(
+                'Outlines does not support function tools, builtin tools or ' + 'output tools yet.'
```
```suggestion
                'Outlines does not support function tools, builtin tools or output tools yet.'
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,352 @@
+                                # The expected format of the assets is not compatible
+                                # between Pydantic-AI and Outlines yet. Outlines will
+                                # widen support in the future.
+                                raise ValueError('Outlines does not support multi-modal ' + 'user prompts yet.')
```
```suggestion
                                raise ValueError('Outlines does not support multi-modal user prompts yet.')
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,352 @@
+                    elif isinstance(part, RetryPromptPart):
+                        chat.add_user_message(str(part.content))
+                    elif isinstance(part, ToolReturnPart):
+                        raise NotImplementedError('Tool calls are not supported for Outlines models ' + 'yet.')
```
```suggestion
                        raise NotImplementedError('Tool calls are not supported for Outlines models yet.')
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,352 @@
+                            BuiltinToolReturnPart,
+                        ),
+                    ):
+                        raise NotImplementedError('Tool calls are not supported for Outlines models ' + 'yet.')
```
```suggestion
                        raise NotImplementedError('Tool calls are not supported for Outlines models yet.')
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,266 @@
+        )
+        model_settings_dict = dict(model_settings) if model_settings else {}
+        if isinstance(self.model, OutlinesAsyncBaseModel):
+            response: str = await self.model(prompt, output_type, None, **model_settings_dict)
```
Copying in what I wrote on Slack:

Another option would be to map the existing fields on ModelSettings  if we know what their equivalent is for the model they're using (e.g. max_tokens , frequency), and tell them to put additional properties in extra_body : https://ai.pydantic.dev/api/settings/#pydantic_ai.settings.ModelSettings.extra_body (which can take an arbitrary dict). I'd like using extra_body better than passing the entire ModelSettings and hitting errors for unsupported keys

RobinPicard commented on tests/models/test_outlines.py:
```diff
@@ -0,0 +1,300 @@
+
+
+async def test_request_streaming(outlines_model: 'Transformers') -> None:
+    # The transformers model does not support streaming,
```
We can test with `llama_cpp`

RobinPicard commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,576 @@
+            settings: Default model settings for this model instance.
+        """
+        self.model: OutlinesBaseModel | OutlinesAsyncBaseModel = model
+        self._model_name: str = 'outlines-model'
```
Not reliably. For some models the user provides a `model` object from the underlying library with a specific model name attached (so not specified to outlines directly), but for others the model name is an optional argument at initialization and can instead only be provided when calling the model.

RobinPicard commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,576 @@
+            raise UserError('Outlines does not support function tools and builtin tools yet.')
+        prompt = self._format_prompt(messages)
+        output_type = (
+            JsonSchema(model_request_parameters.output_object.json_schema)
```
Oh nice! Outlines supports regex and context-free grammar.

RobinPicard commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,576 @@
+    )
+    from outlines.models.vllm_offline import (
+        VLLMOffline,
+        from_vllm_offline,  # pyright: ignore[reportUnknownVariableType]
```
I believe it's because the libraries used in the typing of Outlines are not available in the environment when the linter checks the import.

DouweM commented on docs/index.md:
```diff
@@ -14,7 +14,7 @@ We built Pydantic AI with one simple aim: to bring that FastAPI feeling to GenAI
 
 2. **Model-agnostic**:
-Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral, and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel. If your favorite model or provider is not listed, you can easily implement a [custom model](mo ...
+Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral, Outlines and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel. If your favorite model or provider is not listed, you can easily implement a [custom  ...
```
```suggestion
Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral, and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel, and Outlines. If your favorite model or provider is not listed, you can easily implement a [custom model](models/overview.md#custom-models).
```

DouweM commented on docs/models/outlines.md:
````diff
@@ -0,0 +1,85 @@
+## Install
+
+To use [`OutlinesModel`][pydantic_ai.models.OutlinesModel], you need to either install `pydantic-ai`, or install `pydantic-ai-slim` with the `outlines` optional group:
+```bash
````
This may need another blank line to render correctly

```suggestion

```bash
```

DouweM commented on docs/models/outlines.md:
````diff
@@ -0,0 +1,85 @@
+pip/uv-add "pydantic-ai-slim[outlines]"
+```
+
+As Outlines is a library allowing you to run model from various different providers, it does not include the necessary dependencies for any model, but instead requires you to install the optional group for the provider you want to use. To do so, you can install `pydantic-ai-slim` with an optional group composed of outlines, a dash, and the name of the model. For instance:
````
Does this mean we should drop the `outlines` optional group entirely?

RobinPicard replied:
Ah yes, that makes sense

DouweM commented on docs/models/outlines.md:
```diff
@@ -0,0 +1,85 @@
+
+## Model Initialization
+
+As Outlines is not an inference provider, but instead a library allowing you to run bith local and API-based models, instantiating the model is a bit different from the other models available on Pydantic AI.
```
```suggestion
As Outlines is not an inference provider, but instead a library allowing you to run both local and API-based models, instantiating the model is a bit different from the other models available on Pydantic AI.
```

DouweM commented on docs/models/outlines.md:
```diff
@@ -0,0 +1,85 @@
+
+## Model Settings
+
+As Outlines can be used with various models that do not share the same model settings, the [OutlinesModelSettings][pydantic_ai.models.OutlinesModel.OutlinesModelSettings] class is a bit different from that of the other Pydantic AI models.
```
```suggestion
As Outlines can be used with various models that do not share the same model settings, the [OutlinesModelSettings][pydantic_ai.models.outlines.OutlinesModelSettings] class is a bit different from that of the other Pydantic AI models.
```

DouweM commented on docs/thinking.md:
```diff
@@ -100,3 +100,7 @@ Thinking is supported by the `command-a-reasoning-08-2025` model. It does not ne
+
+## Outlines
+
+Thinking is not supported by Outlines.
```
It's not worth parsing `<think>` tags out of text output, like we do for HuggingFace?

DouweM replied:
Edit: Looks like we already use `split_content_into_text_and_thinking`!

RobinPicard replied:
Ah yes, right!

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,570 @@
+    @classmethod
+    def from_sglang(
+        cls,
+        openai_client: 'openai.OpenAI | openai.AsyncOpenAI',
```
Could it make more sense to provide a base URL and API key instead, and have us build the client from that?

RobinPicard replied:
I see advantages for both options, but I think I prefer the current one to maintain consistency across the loading methods of the `OutlinesModel` class: all of them require the engine/client from the third party library (just like in Outlines). Otherwise I think that could be a bit confusing for users that `from_openai` behaves differently from all others.

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,570 @@
+    @classmethod
+    def from_vllm_offline(  # pragma: no cover
+        cls,
+        vllm_model: Any,
```
```suggestion
        vllm_model: 'vllm.LLM',
```

DouweM commented on docs/models/outlines.md:
```diff
@@ -0,0 +1,85 @@
+- [`from_llamacpp`][pydantic_ai.models.OutlinesModel.from_llamacpp]
+- [`from_mlxlm`][pydantic_ai.models.OutlinesModel.from_mlxlm]
+- [`from_sglang`][pydantic_ai.models.OutlinesModel.from_sglang]
+- [`from_vllm_offline`][pydantic_ai.models.OutlinesModel.from_vllm_offline]
```
I'd like to have an example for each method so it's clear what the user should import and pass.

RobinPicard commented:
I've removed the `outlines` group in the dependencies and changed the `ModelSettings` handling @DouweM. The failing test is unrelated to this PR.

DouweM commented on docs/index.md:
```diff
@@ -14,7 +14,7 @@ We built Pydantic AI with one simple aim: to bring that FastAPI feeling to GenAI
 
 2. **Model-agnostic**:
-Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral, and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel. If your favorite model or provider is not listed, you can easily implement a [custom model](mo ...
+Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Outlines, Mistral and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel. If your favorite model or provider is not listed, you can easily implement a [custom  ...
```
```suggestion
Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel, and Outlines. If your favorite model or provider is not listed, you can easily implement a [custom model](models/overview.md#custom-models).
```

RobinPicard replied:
The comment and suggested change show different things, should Outlines be after Cohere or after Vercel?

DouweM commented on pyproject.toml:
```diff
@@ -46,7 +46,7 @@ requires-python = ">=3.10"
 [tool.hatch.metadata.hooks.uv-dynamic-versioning]
 dependencies = [
-    "pydantic-ai-slim[openai,vertexai,google,groq,anthropic,mistral,cohere,bedrock,huggingface,cli,mcp,evals,ag-ui,retries,temporal,logfire]=={{ version }}",
+    "pydantic-ai-slim[openai,vertexai,google,groq,anthropic,mistral,cohere,bedrock,huggingface,outlines,cli,mcp,evals,ag-ui,retries,temporal,logfire]=={{ version }}",
```
```suggestion
    "pydantic-ai-slim[openai,vertexai,google,groq,anthropic,mistral,cohere,bedrock,huggingface,cli,mcp,evals,ag-ui,retries,temporal,logfire]=={{ version }}",
```

DouweM commented on docs/install.md:
```diff
@@ -51,6 +51,12 @@ pip/uv-add "pydantic-ai-slim[openai]"
 * `cohere` - installs `cohere` [PyPI ↗](https://pypi.org/project/cohere){:target="_blank"}
 * `bedrock` - installs `boto3` [PyPI ↗](https://pypi.org/project/boto3){:target="_blank"}
 * `huggingface` - installs `huggingface-hub[inference]` [PyPI ↗](https://pypi.org/project/huggingface-hub){:target="_blank"}
+* `outlines` - installs `outlines` [PyPI ↗](https://pypi.org/project/outlines){:target="_blank"}
```
```suggestion
```

DouweM commented on docs/models/outlines.md:
```diff
@@ -0,0 +1,134 @@
+
+## Model Settings
+
+As Outlines can be used with various models that do not share the same model settings, the [OutlinesModelSettings][pydantic_ai.models.outlines.OutlinesModelSettings] class is a bit different from that of the other Pydantic AI models.
```
```suggestion
As Outlines can be used with various models that do not share the same model settings, the [`OutlinesModelSettings`][pydantic_ai.models.outlines.OutlinesModelSettings] class is a bit different from that of the other Pydantic AI models.
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,500 @@
+    The docstring of the `ModelSettings` class doscuments the availability of each field. All additional
+    inference arguments should be provided in the `extra_body` dictionary. They will be passed down to
+    the model provider.
+
```
```suggestion
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,500 @@
+    @classmethod
+    def from_vllm_offline(  # pragma: no cover
+        cls,
+        vllm_model: Any,
```
See my other comment, I'd like a more specific type here if possible

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,500 @@
+        """Format the model settings for the inference kwargs."""
+        settings_dict: dict[str, Any] = dict(model_settings) if model_settings else {}
+        extra_body = settings_dict.pop('extra_body', {})
+        settings_dict.update(extra_body)
```
I'd prefer to move this after the formatting of known fields

DouweM commented on docs/thinking.md:
```diff
@@ -100,3 +100,7 @@ Thinking is supported by the `command-a-reasoning-08-2025` model. It does not ne
+
+## Outlines
+
+Some local models run through Outlines include a reasoning part in their output. In that case, it will be handlded by Pydantic AI without the need to specifically enable it. However, Outlines currently does not support thinking along with structured output. If you provide an `output_type`, thinking will be skipped and you may experience degraded performance.
```
```suggestion
Some local models run through Outlines include a reasoning part in their output. In that case, it will be handled by Pydantic AI without the need to specifically enable it. However, Outlines currently does not support thinking along with structured output. If you provide an `output_type`, thinking will be skipped and you may experience degraded performance.
```

We also handle `<think>` tags; we should mention that

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,500 @@
+        """Turn the Outlines text response into a Pydantic AI model response instance."""
+        return ModelResponse(
+            parts=cast(
+                list[ModelResponsePart], split_content_into_text_and_thinking(response, self.profile.thinking_tags)
```
Is "Some local models run through Outlines include a reasoning part in their output. In that case, it will be handlded by Pydantic AI without the need to specifically enable it." in the docs correct? "reasoning part" sounds like it refers to a dedicated reasoning field, not just parsing think tags from text output

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,266 @@
+                    if part.part_kind == 'system-prompt':
+                        chat.add_system_message(part.content)
+                    elif part.part_kind == 'user-prompt':
+                        chat.add_user_message(str(part.content))
```
Not sure how to interpret the downvote 😄 I think this is still todo

RobinPicard commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,494 @@
+        cls,
+        base_url: str,
+        api_key: str | None = None,
+        model_name: str | None = None,
```
No. It's possible to provide the model name during at inference time through the `model` argument in the `extra_body` dict.

DouweM commented on docs/index.md:
```diff
@@ -14,7 +14,7 @@ We built Pydantic AI with one simple aim: to bring that FastAPI feeling to GenAI
 
 2. **Model-agnostic**:
-Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral, and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel. If your favorite model or provider is not listed, you can easily implement a [custom model](mo ...
+Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel, and Outlines. If your favorite model or provider is not listed, you can easily implement a [cus ...
```
```suggestion
Supports virtually every [model](models/overview.md) and provider: OpenAI, Anthropic, Gemini, DeepSeek, Grok, Cohere, Mistral, and Perplexity; Azure AI Foundry, Amazon Bedrock, Google Vertex AI, Ollama, LiteLLM, Groq, OpenRouter, Together AI, Fireworks AI, Cerebras, Hugging Face, GitHub, Heroku, Vercel, and Outlines. If your favorite model or provider is not listed, you can easily implement a [custom model](models/overview.md#custom-models).
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,514 @@
+                    elif isinstance(part, ToolCallPart | BuiltinToolCallPart | BuiltinToolReturnPart):
+                        raise UserError('Tool calls are not supported for Outlines models yet.')
+                    elif isinstance(part, FilePart):
+                        raise UserError('File parts are not supported for Outlines models yet.')
```
Could we include assistant input images (https://ai.pydantic.dev/builtin-tools/#image-generation-tool) as well?

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -413,13 +425,21 @@ def _format_prompt(self, messages: list[ModelMessage]) -> Chat:  # noqa: C901
+                                            image_content['data'], image_content['data_type']
+                                        )
+                                        outlines_input.append(Image(image))
                                     elif isinstance(item, BinaryImage):
```
```suggestion
                                    elif isinstance(item, BinaryContent) and item.is_image:
```

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -433,22 +453,40 @@ def _format_prompt(self, messages: list[ModelMessage]) -> Chat:  # noqa: C901
                         raise UserError('Tool calls are not supported for Outlines models yet.')
                     elif isinstance(part, FilePart):
-                        raise UserError('File parts are not supported for Outlines models yet.')
+                        if isinstance(part.content, BinaryImage):
```
```suggestion
                        if isinstance(item, BinaryContent) and item.is_image:
```

DouweM commented on docs/models/outlines.md:
```diff
@@ -0,0 +1,242 @@
+    items: list[Item]
+    tax: float | None
+    total: float | None
+    date: str | None = Field(pattern=r'\d{4}-\d{2}-\d{2}', description='Date in the format YYYY-MM-DD')
```
This could be `date: date`?

DouweM commented on pydantic_ai_slim/pydantic_ai/_output.py:
```diff
@@ -554,18 +554,12 @@ def __init__(
-        template = self.template or default_template
-
+    @classmethod
+    def create_instructions(cls, template: str, object_def: OutputObjectDefinition) -> str:
```
I'd prefer for the `object_def` to come first

DouweM commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,561 @@
+        """Turn the model messages into an Outlines Chat instance."""
+        chat = Chat()
+
+        if instructions:
```
This reminds me we should support `ModelResponse.instructions` as well. You can use `Model._get_instructions`, there are examples of its usage in other model classes.

DouweM replied:
My bad, I meant `ModelRequest`! But yes that's the logic we need to make sure we don't lose https://ai.pydantic.dev/agents/#instructions

RobinPicard commented on pydantic_ai_slim/pydantic_ai/models/outlines.py:
```diff
@@ -0,0 +1,567 @@
+        if output_format_instructions:
+            chat.add_system_message(output_format_instructions)
+
+        print('HEREEEEE', self._get_instructions(messages))
```
Oops sorry

RobinPicard commented on pydantic_ai_slim/pydantic_ai/_output.py:
```diff
@@ -554,25 +554,29 @@ def __init__(
         return 'prompted'
 
+    @classmethod
+    def create_instructions(cls, template: str, object_def: OutputObjectDefinition) -> str:
```
Sounds good

DouweM commented:
@RobinPicard Heh, we're hitting HuggingFace rate limits:

```
E               429 Client Error: Too Many Requests for url: https://huggingface.co/api/models/erwanf/gpt2-mini/xet-read-token/f12cc7ee54aa4d3e6366597f72ff7acb39b0ab3a (Request ID: Root=1-68fb9343-3b4898b33337e4ed6923263e;4433d677-476f-4296-8e11-8ba47752b357)
E               
E               We had to rate limit your IP (172.183.89.68). To continue using our service, create a HF account or login to your existing account, and make sure you pass a HF_TOKEN if you're using the API.
```

Is there a way to not do any live requests?

RobinPicard commented:
> @RobinPicard Heh, we're hitting HuggingFace rate limits:
> 
> ```
> E               429 Client Error: Too Many Requests for url: https://huggingface.co/api/models/erwanf/gpt2-mini/xet-read-token/f12cc7ee54aa4d3e6366597f72ff7acb39b0ab3a (Request ID: Root=1-68fb9343-3b4898b33337e4ed6923263e;4433d677-476f-4296-8e11-8ba47752b357)
> E               
> E               We had to rate limit your IP (172.183.89.68). To continue using our service, create a HF account or login to your existing account, and make sure you pass a HF_TOKEN if you're using the API.
> ```
> 
> Is there a way to not do any live requests?

Sure but then we would need to mock everything since the error happens during the model loading. Another annoying problem appeared: linting goes through for python3.13 but not for python3.10 because of multiple instances of "Unnecessary "# type: ignore" comment"

RobinPicard commented:
I've made 2 modifications to fix the remaining issues @DouweM 

- Added a commit to add a step in the CI to cache the HF model download. That way models are downloaded only once (instead of over a dozen times) and there's no problem of rate limit anymore. I think this is best as it allows us not to mock anything so that we can have more reliable tests
- Added the `# pyright: reportUnnecessaryTypeIgnoreComment = false` escape command on top of the files `pydantic_ai_slim/pydantic_ai/models/outlines.py` and `tests/outlines/test_outlines.py`. The reason for this is that there was a conflict between the pyright linting for python3.10 and 3.13: one was requesting `pyright: ignore` while the other would fail it were included because of unnecessary ignore commands.
